src/backend/distributed/commands/multi_copy.c

```diff
@@ -0,0 +1,1284 @@
+
+	/* allocate column values and nulls arrays */
+	distributedRelation = heap_open(tableId, RowExclusiveLock);
+	tupleDescriptor = RelationGetDescr(distributedRelation);
+	columnCount = tupleDescriptor->natts;
+	columnValues = palloc0(columnCount * sizeof(Datum));
+	columnNulls = palloc0(columnCount * sizeof(bool));
+
+	/* load the list of shards and verify that we have shards to copy into */
+	shardIntervalList = LoadShardIntervalList(tableId);
+	if (shardIntervalList == NIL)
+	{
+		if (partitionMethod == DISTRIBUTE_BY_HASH)
+		{
+			ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+							errmsg("could not find any shards into which to copy"),
+							errdetail("No shards exist for distributed table \"%s\".",
+									  relationName),
+							errhint("Run master_create_worker_shards to create shards "
+									"and try again.")));
+		}
+		else
+		{
+			ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+							errmsg("could not find any shards into which to copy"),
+							errdetail("No shards exist for distributed table \"%s\".",
+									  relationName)));
+		}
+	}
+
+	/* prevent concurrent placement changes and non-commutative DML statements */
+	LockAllShards(shardIntervalList);
+
+	/* initialize the shard interval cache */
+	shardCount = list_length(shardIntervalList);
+	shardIntervalCache = SortedShardIntervalArray(shardIntervalList);
+
+	/* determine whether to use binary search */
+	if (partitionMethod != DISTRIBUTE_BY_HASH ||
+		!IsUniformHashDistribution(shardIntervalCache, shardCount))
+	{
+		useBinarySearch = true;
+	}
+
+	/* initialize copy state to read from COPY data source */
+	copyState = BeginCopyFrom(distributedRelation,
+							  copyStatement->filename,
+							  copyStatement->is_program,
+							  copyStatement->attlist,
+							  copyStatement->options);
+
+	executorState = CreateExecutorState();
+	executorTupleContext = GetPerTupleMemoryContext(executorState);
+	executorExpressionContext = GetPerTupleExprContext(executorState);
+
+	copyOutState = (CopyOutState) palloc0(sizeof(CopyOutStateData));
+	copyOutState->binary = true;
+	copyOutState->fe_msgbuf = makeStringInfo();
+	copyOutState->rowcontext = executorTupleContext;
+
+	columnOutputFunctions = ColumnOutputFunctions(tupleDescriptor, copyOutState->binary);
+
+	/*
+	 * Create a mapping of shard id to a connection for each of its placements.
+	 * The hash should be initialized before the PG_TRY, since it is used and
+	 * PG_CATCH. Otherwise, it may be undefined in the PG_CATCH (see sigsetjmp
+	 * documentation).
+	 */
+	shardConnectionHash = CreateShardConnectionHash();
+
+	/* set up callback to identify error line number */
+	errorCallback.callback = CopyFromErrorCallback;
+	errorCallback.arg = (void *) copyState;
+	errorCallback.previous = error_context_stack;
+	error_context_stack = &errorCallback;
+
+	/* we use a PG_TRY block to roll back on errors (e.g. in NextCopyFrom) */
+	PG_TRY();
+	{
```

Comment:
> It's actually the other way round imo. Anything that can be touched both
> inside and outside the PG_TRY has to be thought about much more
> careful. PG_TRY/CATCH is implemented using siglongjmp.  

This seems worth documenting. I added a comment on the shardConnectionHash, which is the only variable used in PG_CATCH.  With regards to moving the variables into the PG_TRY block, I'm still concerned that a lot of the calls in the initialization code can produce errors and go into PG_CATCH, which should work out fine, but that wouldn't be immediately clear from the code.